webui/react/src/components/NotebookModal.tsx

```diff
@@ -0,0 +1,309 @@
+import { Button, InputNumber, Modal } from 'antd';
+import { Form, Input, Select } from 'antd';
+import { ModalProps } from 'antd/es/modal/Modal';
+import yaml from 'js-yaml';
+import React, { Dispatch, useCallback, useEffect, useReducer, useState } from 'react';
+
+import useStorage from 'hooks/useStorage';
+import { getResourcePools, getTaskTemplates } from 'services/api';
+import { NotebookConfig, RawJson, ResourcePool, ResourceType, Template } from 'types';
+import { launchNotebook, previewNotebook } from 'utils/task';
+
+import Link from './Link';
+import css from './NotebookModal.module.scss';
+import RadioGroup from './RadioGroup';
+import Spinner from './Spinner';
+
+const MonacoEditor = React.lazy(() => import('react-monaco-editor'));
+
+const { Option } = Select;
+const { Item } = Form;
+
+const STORAGE_PATH = 'notebook-launch';
+const STORAGE_KEY = 'notebook-config';
+
+type DispatchFunction =
+  (Dispatch<{
+    key: keyof NotebookConfig,
+    value: string | number | undefined
+  }>)
+
+function reducer(
+  state: NotebookConfig,
+  action: {key: keyof NotebookConfig, value: string | number | undefined},
+): NotebookConfig {
+  return { ...state, [action.key]: action.value };
+}
+
+const useNotebookForm = (): [NotebookConfig, DispatchFunction] => {
+  const storage = useStorage(STORAGE_PATH);
+  const [ state, dispatch ] = useReducer(
+    reducer,
+    storage.getWithDefault(STORAGE_KEY, { slots: 1 }),
+  );
+
+  const storeConfig = useCallback((values: NotebookConfig) => {
+    delete values.name;
+    storage.set(STORAGE_KEY, values);
+  }, [ storage ]);
+
+  useEffect(() => {
+    storeConfig(state);
+  }, [ state, storeConfig ]);
+
```

Comment:
**suggestion:** Let's leave this for later (after landing this PR).
I think I roughly understand the motivation of wanting to group the form related state code and the related storage code roughly together. However it's fairly small and the reusability aspect of it is not compelling enough to stand on its own. This can probably just sit inside of `NotebookModal` component.